### ISSUE TITLE
unexpected "warn_ungrouped_named_tokens_in_collection" warnings in versions >2.4.0

### ISSUE THREAD

**Issue by a-recknagel** [pyparsing/pyparsing#110]
Hi Paul, I wanted to alert you to this issue in case you didn't hear about it yet: (https://github.com/sdispater/poetry/issues/1244, https://github.com/pypa/packaging/issues/170). The poetry people are working around it by fixing pyparsing to 4.2.0 😕

**Comment by ptmcg** [comment]
Thanks - I just pushed out 2.4.2, which has all these warnings disabled by default.

**Comment by a-recknagel** [comment]
Ah, great. This can be closed then.

**Comment by ptmcg** [comment]
@fungi - thanks for reporting this, though it does sadden me.

I'm confused though - I was under the impression that packaging used a vendored 2.4.7 version of pyparsing, and so was immune to any changes in 3.0.0 until that version got vendored in (after some testing). Can you update this ticket to how these warnings are getting emitted?

**Comment by ptmcg** [comment]
I looked further at the pyparsing code and at your grammar, and this turns out to be a false alarm in pyparsing. I've tightened up the warning detection code, and this will be fixed in 3.0.5.

**Comment by fungi** [comment]
Thanks so much!

**Comment by ptmcg** [comment]
3.0.5 released with this fix

**Comment by fungi** [comment]
Thanks again!

**Comment by ptmcg** [comment]
I believe this fix may have broken this code in requirements.py:

```
MARKER_EXPR.setParseAction(
    lambda s, l, t: Marker(s[t._original_start : t._original_end])
)
```

To fix the warning, I renamed the internally defined results names `_original_start` and `_original_end`. Since MARKER_EXPR is and `originalTextFor` expression, this slicing out of `s` should already be done inside the pyparsing-defined parse action, so should be accessible as `t[0]`.

**Comment by 9999years** [comment]
@ptmcg I'm not super familiar with pyparsing, but that code doesn't seem to work. With `pyparsing` 3.0.5 and `packaging` 20.9, recreating the problematic snippet:

```python
from packaging.markers import MARKER_EXPR, Marker
from pyparsing import ParseResults, originalTextFor


def into_marker(original: str, loc: int, results: ParseResults) -> Marker:
    print("original:", original)
    print("loc:     ", loc)
    print("results: ", repr(results))
    return Marker(results[0])


MARKER_EXPR = originalTextFor(MARKER_EXPR)("marker")
MARKER_EXPR.setParseAction(into_marker)

MARKER_EXPR.parseString('python_version >= "3.6"')
```

Produces the following output:

```
original: python_version >= "3.6"
loc:      0
results:  ParseResults([0, (<Variable('python_version')>, <Op('>=')>, <Value('3.6')>), 23], {'_NOWARN_original_start': 0, '_NOWARN_original_end': 23, 'marker': [0, (<Variable('python_version')>, <Op('>=')>, <Value('3.6')>), 23]})
Traceback (most recent call last):
  File "xxx_parse_reqs.py", line 40, in <module>
    print(MARKER_EXPR.parseString('python_version >= "3.6"'))
  File "/home/rturner/.cache/pypoetry/virtualenvs/example-package-JfO58x2Y-py3.7/lib/python3.7/site-packages/pyparsing/core.py", line 1100, in parse_string
    loc, tokens = self._parse(instring, 0)
  File "/home/rturner/.cache/pypoetry/virtualenvs/example-package-JfO58x2Y-py3.7/lib/python3.7/site-packages/pyparsing/core.py", line 827, in _parseNoCache
    tokens = fn(instring, tokens_start, ret_tokens)
  File "/home/rturner/.cache/pypoetry/virtualenvs/example-package-JfO58x2Y-py3.7/lib/python3.7/site-packages/pyparsing/core.py", line 283, in wrapper
    ret = func(*args[limit:])
  File "xxx_parse_reqs.py", line 34, in into_marker
    return Marker(results[0])
  File "/home/rturner/.cache/pypoetry/virtualenvs/example-package-JfO58x2Y-py3.7/lib/python3.7/site-packages/packaging/markers.py", line 307, in __init__
    self._markers = _coerce_parse_result(MARKER.parseString(marker))
  File "/home/rturner/.cache/pypoetry/virtualenvs/example-package-JfO58x2Y-py3.7/lib/python3.7/site-packages/pyparsing/core.py", line 1098, in parse_string
    instring = instring.expandtabs()
AttributeError: 'int' object has no attribute 'expandtabs'
```

It doesn't seem to change depending on the usage of the `as_string` argument to `original_text_for`, either.

**Comment by ptmcg** [comment]
In your code that doesn't work, this line:

    MARKER_EXPR.setParseAction(into_marker)

should be:

    MARKER_EXPR.addParseAction(into_marker)

**Comment by 9999years** [comment]
Ah, thank you! That change works -- I'm not sure why the downstream was using `setParseAction` instead of `addParseAction`.

**Comment by ptmcg** [comment]
It is very old code, and probably predates when `addParseAction` was added to pyparsing.

**Comment by ptmcg** [comment]
Fixed (for the last time!) in 3.0.6
